onnxruntime/core/providers/webgpu/tensor/gather.cc

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/webgpu/tensor/gather.h"
+#include "core/providers/webgpu/shader_helper.h"
+#include "core/providers/webgpu/webgpu_supported_types.h"
+
+namespace onnxruntime {
+namespace webgpu {
+
+Status GatherProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  const auto& data = shader.AddInput("data", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);
+  const auto& indices = shader.AddInput("input_indices", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias | ShaderUsage::UseValueTypeAlias);
+  const auto& output = shader.AddOutput("output", ShaderUsage::UseUniform);
+
+  std::ostringstream calc_data_indices;
+  calc_data_indices.imbue(std::locale::classic());
+  calc_data_indices << "var indices_indices  = input_indices_indices_t(0);\n";
+  for (int i = 0; i < indices.Rank(); i++) {
+    calc_data_indices << "indices_indices";
+    if (indices.Rank() > 1) {
+      calc_data_indices << "[" << i << "]";
+    }
```

Comment:
maybe can use AddIncides to add a indices helper for "indices_indices" (oh this name is confusing...) and use `IndicesGet`/`IndicesSet` so that no need to use the `Rank()`